# Student Do: Three-Year Loans

The higher-ups at Harold's firm want to know how to better target customers who are seeking three-year loans. Harold's manager has asked him to review a compilation of loan data, and then filter out the necessary data to generate insights about customers who have been granted three-year loans.

Follow the instructions to help Harold answer his manager's questions.

## Instructions

Complete the following steps.

### Import the necessary libraries and dependencies.

In [68]:
# Import libraries and dependencies
from pathlib import Path
import csv
import pandas as pd

### Read CSV in as DataFrame

In [69]:
# Set the path
data = Path("../Resources/loans.csv")

# Read in the CSV as a DataFrame
data_df = pd.read_csv(data)
data_df.head()

C:\Users\TribThapa\Anaconda3\lib\site-packages\IPython\core\interactiveshell.py:3020: DtypeWarning: Columns (11,17) have mixed types.Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


loan_amnt  funded_amnt  funded_amnt_inv       term  int_rate  installment  \
0      10000        10000          10000.0  36 months    0.1033       324.23   
1       4000         4000           4000.0  36 months    0.2340       155.68   
2       5000         5000           5000.0  36 months    0.1797       180.69   
3       9600         9600           9600.0  36 months    0.1298       323.37   
4       2500         2500           2500.0  36 months    0.1356        84.92   

  grade sub_grade       emp_title emp_length home_ownership annual_inc  \
0     B        B1             NaN   < 1 year       MORTGAGE     280000   
1     E        E1        Security    3 years           RENT      90000   
2     D        D1  Administrative    6 years       MORTGAGE      59280   
3     B        B5             NaN        NaN       MORTGAGE      35704   
4     C        C1            Chef  10+ years           RENT      55000   

  verification_status issue_d loan_status pymnt_plan             purpose  \
0        Not Verified  Dec-18     Current          n  debt_consolidation   
1     Source Verified  Dec-18     Current          n  debt_consolidation   
2     Source Verified  Dec-18     Current          n  debt_consolidation   
3        Not Verified  Dec-18     Current          n    home_improvement   
4        Not Verified  Dec-18     Current          n  debt_consolidation   

  Unnamed: 17  
0         NaN  
1         NaN  
2         NaN  
3         NaN  
4         NaN

### Show the First 10 Records

In [70]:
# Retrieve rows with index 0 up to 10 (not including)
data_df.head(10)

loan_amnt  funded_amnt  funded_amnt_inv       term  int_rate  installment  \
0      10000        10000          10000.0  36 months    0.1033       324.23   
1       4000         4000           4000.0  36 months    0.2340       155.68   
2       5000         5000           5000.0  36 months    0.1797       180.69   
3       9600         9600           9600.0  36 months    0.1298       323.37   
4       2500         2500           2500.0  36 months    0.1356        84.92   
5      30000        30000          30000.0  60 months    0.1894       777.23   
6      23000        23000          23000.0  60 months    0.2089       620.81   
7      32075        32075          32075.0  60 months    0.1180       710.26   
8       8000         8000           8000.0  36 months    0.2340       311.35   
9      10000        10000          10000.0  60 months    0.1992       264.50   

  grade sub_grade           emp_title emp_length home_ownership annual_inc  \
0     B        B1                 NaN   < 1 year       MORTGAGE     280000   
1     E        E1            Security    3 years           RENT      90000   
2     D        D1      Administrative    6 years       MORTGAGE      59280   
3     B        B5                 NaN        NaN       MORTGAGE      35704   
4     C        C1                Chef  10+ years           RENT      55000   
5     D        D2         Postmaster   10+ years       MORTGAGE      90000   
6     D        D4            Operator    5 years           RENT      68107   
7     B        B4  Nursing Supervisor  10+ years       MORTGAGE     150000   
8     E        E1             Manager  10+ years            OWN      43000   
9     D        D3    Material Handler  10+ years       MORTGAGE      80000   

  verification_status issue_d loan_status pymnt_plan             purpose  \
0        Not Verified  Dec-18     Current          n  debt_consolidation   
1     Source Verified  Dec-18     Current          n  debt_consolidation   
2     Source Verified  Dec-18     Current          n  debt_consolidation   
3        Not Verified  Dec-18     Current          n    home_improvement   
4        Not Verified  Dec-18     Current          n  debt_consolidation   
5     Source Verified  Dec-18     Current          n  debt_consolidation   
6     Source Verified  Dec-18     Current          n  debt_consolidation   
7        Not Verified  Dec-18     Current          n         credit_card   
8     Source Verified  Dec-18     Current          n  debt_consolidation   
9        Not Verified  Dec-18     Current          n  debt_consolidation   

  Unnamed: 17  
0         NaN  
1         NaN  
2         NaN  
3         NaN  
4         NaN  
5         NaN  
6         NaN  
7         NaN  
8         NaN  
9         NaN

### View Summary Statistics

In [71]:
# Describe summary statistics for csv data
data_df.describe()

loan_amnt    funded_amnt  funded_amnt_inv       int_rate  \
count  128412.000000  128412.000000    128412.000000  128412.000000   
mean    15971.321021   15971.321021     15968.498166       0.129302   
std     10150.384233   10150.384233     10152.168970       0.050904   
min      1000.000000    1000.000000       725.000000       0.060000   
25%      8000.000000    8000.000000      8000.000000       0.084600   
50%     14000.000000   14000.000000     14000.000000       0.118000   
75%     21600.000000   21600.000000     21600.000000       0.161400   
max     40000.000000   40000.000000     40000.000000       0.309900   

         installment  
count  128412.000000  
mean      463.259353  
std       285.718256  
min        30.480000  
25%       253.630000  
50%       382.905000  
75%       622.680000  
max      1618.240000

In [72]:
data_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 128412 entries, 0 to 128411
Data columns (total 18 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   loan_amnt            128412 non-null  int64  
 1   funded_amnt          128412 non-null  int64  
 2   funded_amnt_inv      128412 non-null  float64
 3   term                 128412 non-null  object 
 4   int_rate             128412 non-null  float64
 5   installment          128412 non-null  float64
 6   grade                128412 non-null  object 
 7   sub_grade            128412 non-null  object 
 8   emp_title            107465 non-null  object 
 9   emp_length           116708 non-null  object 
 10  home_ownership       128412 non-null  object 
 11  annual_inc           128412 non-null  object 
 12  verification_status  128412 non-null  object 
 13  issue_d              128412 non-null  object 
 14  loan_status          128412 non-null  object 
 15  pymnt_plan       

### Filter Columns and Create Subset DataFrame

In [73]:
# Filter the DataFrame down to the following, keep all rows

col = ['loan_amnt', 'term', 'int_rate', 'emp_title', 'annual_inc', 'purpose']

data_df_filtered = data_df.loc[:, col]
data_df_filtered

#loan_amnt
#term
#int_rate
#emp_title
#annual_inc
#purpose


loan_amnt       term  int_rate             emp_title annual_inc  \
0           10000  36 months    0.1033                   NaN     280000   
1            4000  36 months    0.2340              Security      90000   
2            5000  36 months    0.1797        Administrative      59280   
3            9600  36 months    0.1298                   NaN      35704   
4            2500  36 months    0.1356                  Chef      55000   
...           ...        ...       ...                   ...        ...   
128407      23000  36 months    0.1502        Tax Consultant    75000.0   
128408      10000  36 months    0.1502        security guard    38000.0   
128409       5000  36 months    0.1356          Payoff Clerk    35360.0   
128410      10000  36 months    0.1106                   NaN    44400.0   
128411      10000  36 months    0.1691  Key Accounts Manager    80000.0   

                   purpose  
0       debt_consolidation  
1       debt_consolidation  
2       debt_consolidation  
3         home_improvement  
4       debt_consolidation  
...                    ...  
128407  debt_consolidation  
128408  debt_consolidation  
128409  debt_consolidation  
128410         credit_card  
128411               other  

[128412 rows x 6 columns]

### Filter Rows by 36 Month Terms

In [74]:
# Conditional indexing to filter DataFrame where 'term' is equal to '36 months'
data_df_filtered_months = data_df_filtered.loc[data_df_filtered['term'] == '36 months']
data_df_filtered_months

loan_amnt       term  int_rate             emp_title annual_inc  \
0           10000  36 months    0.1033                   NaN     280000   
1            4000  36 months    0.2340              Security      90000   
2            5000  36 months    0.1797        Administrative      59280   
3            9600  36 months    0.1298                   NaN      35704   
4            2500  36 months    0.1356                  Chef      55000   
...           ...        ...       ...                   ...        ...   
128407      23000  36 months    0.1502        Tax Consultant    75000.0   
128408      10000  36 months    0.1502        security guard    38000.0   
128409       5000  36 months    0.1356          Payoff Clerk    35360.0   
128410      10000  36 months    0.1106                   NaN    44400.0   
128411      10000  36 months    0.1691  Key Accounts Manager    80000.0   

                   purpose  
0       debt_consolidation  
1       debt_consolidation  
2       debt_consolidation  
3         home_improvement  
4       debt_consolidation  
...                    ...  
128407  debt_consolidation  
128408  debt_consolidation  
128409  debt_consolidation  
128410         credit_card  
128411               other  

[88179 rows x 6 columns]

### Modify Row Values

In [75]:
# Change row values within the 'term' column from '36 months' to '3 Years'
data_df_filtered_months.iloc[:, 1] = '3 Years'
data_df_filtered_months

C:\Users\TribThapa\Anaconda3\lib\site-packages\pandas\core\indexing.py:1720: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self._setitem_single_column(loc, value, pi)


loan_amnt     term  int_rate             emp_title annual_inc  \
0           10000  3 Years    0.1033                   NaN     280000   
1            4000  3 Years    0.2340              Security      90000   
2            5000  3 Years    0.1797        Administrative      59280   
3            9600  3 Years    0.1298                   NaN      35704   
4            2500  3 Years    0.1356                  Chef      55000   
...           ...      ...       ...                   ...        ...   
128407      23000  3 Years    0.1502        Tax Consultant    75000.0   
128408      10000  3 Years    0.1502        security guard    38000.0   
128409       5000  3 Years    0.1356          Payoff Clerk    35360.0   
128410      10000  3 Years    0.1106                   NaN    44400.0   
128411      10000  3 Years    0.1691  Key Accounts Manager    80000.0   

                   purpose  
0       debt_consolidation  
1       debt_consolidation  
2       debt_consolidation  
3         home_improvement  
4       debt_consolidation  
...                    ...  
128407  debt_consolidation  
128408  debt_consolidation  
128409  debt_consolidation  
128410         credit_card  
128411               other  

[88179 rows x 6 columns]

In [76]:
# Change row values within the 'emp_title' column from NaN to 'Unknown'
data_df_filtered_months['emp_title'].fillna('Unknown', inplace=True)
data_df_filtered_months

C:\Users\TribThapa\Anaconda3\lib\site-packages\pandas\core\series.py:4469: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  downcast=downcast,


loan_amnt     term  int_rate             emp_title annual_inc  \
0           10000  3 Years    0.1033               Unknown     280000   
1            4000  3 Years    0.2340              Security      90000   
2            5000  3 Years    0.1797        Administrative      59280   
3            9600  3 Years    0.1298               Unknown      35704   
4            2500  3 Years    0.1356                  Chef      55000   
...           ...      ...       ...                   ...        ...   
128407      23000  3 Years    0.1502        Tax Consultant    75000.0   
128408      10000  3 Years    0.1502        security guard    38000.0   
128409       5000  3 Years    0.1356          Payoff Clerk    35360.0   
128410      10000  3 Years    0.1106               Unknown    44400.0   
128411      10000  3 Years    0.1691  Key Accounts Manager    80000.0   

                   purpose  
0       debt_consolidation  
1       debt_consolidation  
2       debt_consolidation  
3         home_improvement  
4       debt_consolidation  
...                    ...  
128407  debt_consolidation  
128408  debt_consolidation  
128409  debt_consolidation  
128410         credit_card  
128411               other  

[88179 rows x 6 columns]

In [ ]:
data_df_filtered_months['is_Null'] = data_df_filtered_months['emp_title'].isnull()

data_df_filtered_months.loc[data_df_filtered_months.is_Null == True, 'emp_title'] = "Unknown"

data_df_filtered_months.head()

### View Summary Statistics for 3 Year Term Loans

In [77]:
# Describe summary statistics for three-year loans
data_df_filtered_months.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 88179 entries, 0 to 128411
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   loan_amnt   88179 non-null  int64  
 1   term        88179 non-null  object 
 2   int_rate    88179 non-null  float64
 3   emp_title   88179 non-null  object 
 4   annual_inc  88179 non-null  object 
 5   purpose     88179 non-null  object 
dtypes: float64(1), int64(1), object(4)
memory usage: 4.7+ MB


In [78]:
data_df_filtered_months.describe()

loan_amnt      int_rate
count  88179.000000  88179.000000
mean   13279.580456      0.120610
std     9479.610905      0.048955
min     1000.000000      0.060000
25%     6000.000000      0.078400
50%    10000.000000      0.110600
75%    18000.000000      0.150200
max    40000.000000      0.309900

### Show Unique Values for Employee Titles with 3 Year Loans

In [79]:
# Calculate unique values and counts for employee titles of 3 year customer loans
data_df_filtered_months['emp_title'].nunique()

31461

In [80]:
data_df_filtered_months['emp_title'].count()

88179

In [81]:
data_df_filtered_months['emp_title'].agg(['nunique', 'count'])

nunique    31461
count      88179
Name: emp_title, dtype: int64

In [90]:
data_df_filtered_months['emp_title'].value_counts()

Unknown                       15669
Teacher                        1433
Manager                        1207
Registered Nurse                672
Driver                          654
                              ...  
Contrtoller                       1
Service plumber                   1
Packaging Specialist              1
director's admin assistant        1
Police officers                   1
Name: emp_title, Length: 31461, dtype: int64

In [91]:
data_df_filtered_months['purpose'].value_counts()

debt_consolidation    45452
credit_card           24779
other                  5836
home_improvement       5446
major_purchase         1795
medical                1245
car                     838
small_business          828
vacation                749
house                   596
moving                  547
renewable_energy         63
n                         5
Name: purpose, dtype: int64

### Show Unique Values for Purpose of 3 Year Loans

In [82]:
# Calculate unique values and counts for loan purposes of 3 year customer loans
data_df_filtered_months['purpose'].agg(['nunique', 'count'])

nunique       13
count      88179
Name: purpose, dtype: int64

### Show Summary Statistics for 3 Year Loans of Customers with Annual Income Greater Than $80,000

In [83]:
# Display summary statistics where annual income is greater than $80,000 to find count and mean

data_df_filtered_months['annual_inc'] = data_df_filtered_months['annual_inc'].replace('RENT', '0')
data_df_filtered_months['annual_inc'] = data_df_filtered_months['annual_inc'].replace('MORTGAGE', '0')

data_df_filtered_months['annual_inc'] = data_df_filtered_months["annual_inc"].astype('float')

data_df_filtered_months_80000 = data_df_filtered_months.loc[data_df_filtered_months['annual_inc'] > 80000]

# Display summary statistics where annual income is less than $80,000 to find count and mean
data_df_filtered_months_80000.describe()


C:\Users\TribThapa\Anaconda3\lib\site-packages\ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  This is separate from the ipykernel package so we can avoid doing imports until
C:\Users\TribThapa\Anaconda3\lib\site-packages\ipykernel_launcher.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  after removing the cwd from sys.path.
C:\Users\TribThapa\Anaconda3\lib\site-packages\ipykernel_launcher.py:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataF

loan_amnt      int_rate    annual_inc
count  29746.000000  29746.000000  2.974600e+04
mean   18240.233141      0.114118  1.389637e+05
std    10992.262506      0.048068  1.752417e+05
min     1000.000000      0.060000  8.000100e+04
25%    10000.000000      0.075600  9.500000e+04
50%    15700.000000      0.104700  1.120000e+05
75%    25000.000000      0.144700  1.500000e+05
max    40000.000000      0.309400  9.000000e+06

In [84]:
data_df_filtered_months_80000.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 29746 entries, 0 to 128393
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   loan_amnt   29746 non-null  int64  
 1   term        29746 non-null  object 
 2   int_rate    29746 non-null  float64
 3   emp_title   29746 non-null  object 
 4   annual_inc  29746 non-null  float64
 5   purpose     29746 non-null  object 
dtypes: float64(2), int64(1), object(3)
memory usage: 1.6+ MB


### Show Summary Statistics for 3 Year Loans of Customers with Annual Income Less Than $80,000

In [86]:
data_df_filtered_months_less80000 = data_df_filtered_months.loc[data_df_filtered_months['annual_inc'] < 80000]

data_df_filtered_months_less80000.describe()

loan_amnt      int_rate    annual_inc
count  56017.000000  56017.000000  56017.000000
mean   10607.698110      0.124267  49367.794171
std     7318.751423      0.049136  16674.943090
min     1000.000000      0.060000      0.000000
25%     5000.000000      0.081900  37000.000000
50%     9600.000000      0.113100  50000.000000
75%    14000.000000      0.150200  62605.870000
max    40000.000000      0.309900  79999.000000

In [87]:
data_df_filtered_months_less80000.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 56017 entries, 2 to 128410
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   loan_amnt   56017 non-null  int64  
 1   term        56017 non-null  object 
 2   int_rate    56017 non-null  float64
 3   emp_title   56017 non-null  object 
 4   annual_inc  56017 non-null  float64
 5   purpose     56017 non-null  object 
dtypes: float64(2), int64(1), object(3)
memory usage: 3.0+ MB


### What insights can we answer regarding 3 year loans?

1. What kind of customers (employee title) seem to ask for three-year loans most frequently? **Teacher: 1433**
2. What are three-year loans generally used for? **Debt consolidation 45452**
3. What is the difference in count of three-year loan customers with annual incomes greater than 80,000 compared to those with annual incomes less than 80,000? **29746 - 56017 = -26271**
4. What is the difference in interest rates of three-year loan customers with annual incomes greater than 80,000 compared to those with annual incomes less than 80,000? **0.114118 - 0.124267 = -0.010149**

In [ ]:
|